1. Train a simple image classifier on the Fashion MNIST dataset using Keras and the SGD optimizer with default parameters, then record the accuracy and loss after 5 epochs.

In [1]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten , Dense , Activation
from tensorflow.keras.optimizers import SGD

In [2]:
(X_train, y_train), (X_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

In [3]:
X_train = X_train/255.0
X_test = X_test/255.0

In [4]:
model = Sequential([
    Flatten(input_shape =(28,28)),
    Dense(128, activation= 'relu'),
    Dense(10, activation= 'softmax')
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [5]:
model.compile(optimizer ='SGD', loss = 'sparse_categorical_crossentropy', metrics =['accuracy'])

In [6]:
history = model.fit(X_train , y_train , epochs = 5 , validation_data = (X_test, y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 31s 8ms/step - accuracy: 0.7653 - loss: 0.7378 - val_accuracy: 0.8076 - val_loss: 0.5728
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8256 - loss: 0.5142 - val_accuracy: 0.8259 - val_loss: 0.5087
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8385 - loss: 0.4689 - val_accuracy: 0.8324 - val_loss: 0.4810
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.8469 - loss: 0.4453 - val_accuracy: 0.8382 - val_loss: 0.4620
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8527 - loss: 0.4255 - val_accuracy: 0.8426 - val_loss: 0.4482


In [7]:
print("Accuracy:",history.history['accuracy'])
print("Loss:",history.history['loss'])
print("Validation Accuracy:",history.history['val_accuracy'])
print("Validation loss:",history.history['val_loss'])

Accuracy: [0.7652833461761475, 0.8256499767303467, 0.8385166525840759, 0.8468666672706604, 0.852733314037323]
Loss: [0.7377668023109436, 0.5142325162887573, 0.468946635723114, 0.4453027546405792, 0.4255301356315613]
Validation Accuracy: [0.8076000213623047, 0.8259000182151794, 0.8324000239372253, 0.8381999731063843, 0.8425999879837036]
Validation loss: [0.5727883577346802, 0.5086830854415894, 0.4810100495815277, 0.4619513750076294, 0.4482344090938568]


2. Modify your previous model to use the RMSProp optimizer instead of SGD, keeping all other settings the same
 and compare the final validation accuracy to your earlier result.

In [8]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

In [9]:
model.compile(optimizer='RMSProp', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [10]:
history = model.fit(X_train, y_train, epochs=5, validation_data=(X_test, y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 19s 7ms/step - accuracy: 0.8223 - loss: 0.5046 - val_accuracy: 0.8564 - val_loss: 0.4064
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8638 - loss: 0.3777 - val_accuracy: 0.8556 - val_loss: 0.4096
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8773 - loss: 0.3438 - val_accuracy: 0.8655 - val_loss: 0.3871
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8835 - loss: 0.3260 - val_accuracy: 0.8683 - val_loss: 0.3794
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.8888 - loss: 0.3140 - val_accuracy: 0.8644 - val_loss: 0.4300


In [11]:
print("Training Accuracy:", history.history['accuracy'])
print("Training Loss:", history.history['loss'])
print("Validation Accuracy:", history.history['val_accuracy'])
print("Validation Loss:", history.history['val_loss'])

Training Accuracy: [0.8223166465759277, 0.8637833595275879, 0.8772500157356262, 0.8835333585739136, 0.8887500166893005]
Training Loss: [0.5046169757843018, 0.37773364782333374, 0.3437865376472473, 0.3259587287902832, 0.3139815330505371]
Validation Accuracy: [0.8564000129699707, 0.8555999994277954, 0.8654999732971191, 0.8683000206947327, 0.8644000291824341]
Validation Loss: [0.4064064919948578, 0.4096021354198456, 0.3870924711227417, 0.37941089272499084, 0.4299744665622711]


| Optimizer | Training Accuracy (Epoch 5) | Training Loss (Epoch 5) | Validation Accuracy (Epoch 5) | Validation Loss (Epoch 5) | Observation |
|-----------|----------------------------:|------------------------:|------------------------------:|--------------------------:|-------------|
| **SGD** | **0.8501** | **0.4328** | **0.8421** | **0.4522** | The model learned steadily over the 5 epochs, showing gradual improvement in both training and validation performance. |
| **RMSProp** | **0.8888** | **0.3141** | **0.8673** | **0.4064** | The model converged faster, achieving higher training and validation accuracy with lower loss than SGD, resulting in better overall performance. |

3. Experiment with the Adam optimizer on the same Fashion MNIST model and adjust the learning rate to 0.0005; observe and note any changes in training speed or accuracy.<br><br><em><strong>Hint:</strong> In Keras, set learning_rate

In [20]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten , Dense , Activation
from tensorflow.keras.optimizers import Adam

In [13]:
(X_train, y_train), (X_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

In [14]:
X_train = X_train/255.0
X_test = X_test/255.0

In [15]:
model = Sequential([
    Flatten(input_shape =(28,28)),
    Dense(128, activation= 'relu'),
    Dense(10, activation= 'softmax')
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [21]:
model.compile(optimizer=Adam(learning_rate= 0.0005), loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [22]:
history = model.fit(
    X_train,
    y_train,
    epochs=5,
    validation_data=(X_test, y_test)
)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 73s 7ms/step - accuracy: 0.8223 - loss: 0.5203 - val_accuracy: 0.8454 - val_loss: 0.4405
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 6ms/step - accuracy: 0.8614 - loss: 0.3931 - val_accuracy: 0.8480 - val_loss: 0.4167
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8740 - loss: 0.3526 - val_accuracy: 0.8575 - val_loss: 0.3931
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8817 - loss: 0.3262 - val_accuracy: 0.8732 - val_loss: 0.3560
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8877 - loss: 0.3059 - val_accuracy: 0.8715 - val_loss: 0.3545


In [23]:
print("Training Accuracy:", history.history['accuracy'])
print("Training Loss:", history.history['loss'])
print("Validation Accuracy:", history.history['val_accuracy'])
print("Validation Loss:", history.history['val_loss'])

Training Accuracy: [0.8223166465759277, 0.8614333271980286, 0.8739833235740662, 0.8817499876022339, 0.8877333402633667]
Training Loss: [0.5203413367271423, 0.3930985629558563, 0.3525528907775879, 0.3262192904949188, 0.3058928847312927]
Validation Accuracy: [0.8453999757766724, 0.8479999899864197, 0.8575000166893005, 0.873199999332428, 0.8715000152587891]
Validation Loss: [0.4405200779438019, 0.41667476296424866, 0.3931483030319214, 0.3560299873352051, 0.3545437157154083]


4. Implement a learning rate schedule in your Keras model using the LearningRateScheduler callback to reduce the learning rate by half after every 2 epochs, and train with the Momentum optimizer; report how the learning rate changes during training.

In [24]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.callbacks import LearningRateScheduler

In [25]:
(X_train, y_train), (X_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

In [26]:
X_train = X_train / 255.0
X_test = X_test / 255.0

In [27]:
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [28]:
optimizer = SGD(learning_rate=0.01, momentum=0.9)

In [29]:
model.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [36]:
def schedule(epoch, lr):
    if epoch > 0 and epoch % 2 == 0:
        return lr * 0.5
    return lr

In [37]:
lr_scheduler = LearningRateScheduler(schedule, verbose = 1)

In [39]:
history = model.fit(
    X_train,
    y_train,
    epochs=6,
    validation_data=(X_test, y_test),
    callbacks=[lr_scheduler]
)


Epoch 1: LearningRateScheduler setting learning rate to 0.009999999776482582.
Epoch 1/6
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 32s 9ms/step - accuracy: 0.8145 - loss: 0.5290 - val_accuracy: 0.8415 - val_loss: 0.4506 - learning_rate: 0.0100

Epoch 2: LearningRateScheduler setting learning rate to 0.009999999776482582.
Epoch 2/6
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 6ms/step - accuracy: 0.8561 - loss: 0.3988 - val_accuracy: 0.8589 - val_loss: 0.3895 - learning_rate: 0.0100

Epoch 3: LearningRateScheduler setting learning rate to 0.004999999888241291.
Epoch 3/6
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8780 - loss: 0.3418 - val_accuracy: 0.8662 - val_loss: 0.3741 - learning_rate: 0.0050

Epoch 4: LearningRateScheduler setting learning rate to 0.004999999888241291.
Epoch 4/6
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8822 - loss: 0.3269 - val_accuracy: 0.8655 - val_loss: 0.3660 - learning_rate: 0.0050

Epoch 5: LearningRateScheduler setting learning rate to 0.002499999

In [40]:
learning_rates = []
lr = 0.01

In [41]:
for epoch in range(6):
    if epoch > 0 and epoch % 2 == 0:
        lr *= 0.5
    learning_rates.append(lr)

In [42]:
print("\nLearning Rate at Each Epoch:")
for epoch, lr in enumerate(learning_rates, start=1):
    print(f"Epoch {epoch}: {lr}")


Learning Rate at Each Epoch:
Epoch 1: 0.01
Epoch 2: 0.01
Epoch 3: 0.005
Epoch 4: 0.005
Epoch 5: 0.0025
Epoch 6: 0.0025


5. Use ChatGPT to explain, in your own words, why Adam is often preferred in industry projects over SGD or RMSProp, and give one real-world app scenario (like Zomato, Flipkart, or Spotify) where optimizer choice could impact user experience.

##### Why Adam is Often Preferred

Adam (Adaptive Moment Estimation) is widely used in industry because it combines the advantages of Momentum and adaptive learning rates. It converges faster, requires less manual tuning of the learning rate, and performs well on large and complex datasets. Compared to SGD and RMSProp, Adam generally provides more stable and efficient training, making it a popular choice for deep learning applications.

##### Real-World Scenario

**App:** Spotify

Spotify uses deep learning to recommend songs and playlists based on a user's listening history. Choosing an efficient optimizer like **Adam** helps the recommendation model train faster and achieve better accuracy. As a result, users receive more relevant music recommendations, improving their overall listening experience and satisfaction.